metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -582,16 +582,17 @@ ReplicationControlManager replicationControl() {
         return replicationControl;
     }
 
-    // VisibleForTesting
-    <T> CompletableFuture<T> appendReadEvent(String name, Supplier<T> handler) {
-        ControllerReadEvent<T> event = new ControllerReadEvent<T>(name, handler);
-        queue.append(event);
-        return event.future();
-    }
-
-    <T> CompletableFuture<T> appendReadEvent(String name, long deadlineNs, Supplier<T> handler) {
+    <T> CompletableFuture<T> appendReadEvent(
+        String name,
+        OptionalLong deadlineNs,
+        Supplier<T> handler
+    ) {
         ControllerReadEvent<T> event = new ControllerReadEvent<T>(name, handler);
-        queue.appendWithDeadline(deadlineNs, event);
+        if (deadlineNs.isPresent()) {
+            queue.appendWithDeadline(deadlineNs.getAsLong(), event);
+        } else {
+            queue.append(event);
+        }
```

Comment:
Could you provide some context here. Is this related to the issue with fairness we saw for expired events?

Reply:
Some controller operations have a deadline, after which they should just be timed out rather than executed. This is important because the client may have retried the operation in the meantime. So keeping all the old operations can lead to a kind of congestion collapse.

We probably should set a deadline for all operations, but that would be a bigger change. This change just puts the timeout in the context object rather than passing it as a separate parameter.